### PR TITLE
Add the code used in Spring 2024 bootcamp

### PR DESCRIPTION
**Pull request by lanlou1554**
Add the code used in Spring 2024 bootcamp.
The code dives deeply into C++ new features like move constructor/assignment, move semantics, unique_ptr, shared_ptr, wrapper class, etc., by implementing a simple version of unique_ptr from scratch.

(This PR also makes wrapper_class.cpp safer)

### CODE REVIEW

File: src/spring2024/my_ptr.cpp [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+  // Copy-init constructor is explicitely deleted.
+  Pointer(const Pointer<T> &) = delete;
+  // Copy-assign operator is explicitely deleted.
+  Pointer<T> &operator=(const Pointer<T> &) = delete;
+
+  // Add move constructor: useful when we need to EXTEND the lifetime of an object!
+  Pointer<T>(Pointer<T> &&another) : ptr_(another.ptr_) { another.ptr_ = nullptr; }
+  // Add move assign operator: useful when we need to EXTEND the lifetime of an object!
+  Pointer<T> &operator=(Pointer<T> &&another) {
```
Comment (skyzh) [inline src/spring2024/my_ptr.cpp:36]: need check this == another

File: src/spring2024/my_ptr.cpp [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+// wrapper class, etc., by implementing a simple version of unique_ptr from scratch.
+// **IMPORTANT**: please BEGIN reading from MAIN function!
+
+// Notice: it is just a simple version of std::unique_pointer<T>. The real implementation is much more complex!
+// A template allows us to replace any type T, with what we want later in our code.
+template <typename T>
+class Pointer {
+ public:
+  Pointer(T val) {
+    ptr_ = new T;
```
Comment (skyzh) [inline src/spring2024/my_ptr.cpp:16]: does not work for types w/o default constructor, but it's fine for educational purposes...

File: src/spring2024/my_ptr.cpp (hunk 1 above)
Comment (lanlou1554) [reply src/spring2024/my_ptr.cpp]: Updated! Good Catch!

File: src/spring2024/my_ptr.cpp (hunk 2 above)
Comment (lanlou1554) [reply src/spring2024/my_ptr.cpp]: Updated. Thx!

File: README.md [hunk 3]
```diff
@@ -62,6 +62,9 @@ reading the files on concepts you are unfamiliar about.
 - `condition_variable.cpp`: Covers `std::condition_variable`.
 - `rwlock.cpp`: Covers the usage of several C++ STL synchronization primitive libraries (`std::shared_mutex`, `std::shared_lock`, `std::unique_lock`) to create a reader-writer's lock implementation. 
 
+### Demo Codes for 445/645 Bootcamp
```
Comment (abigalekim) [inline README.md:65]: NIT: Demo Code for 15-445/645 Bootcamp
and likewise, "Covers the code used in ... "

File: src/spring2024/s24_my_ptr.cpp [hunk 4]
```diff
@@ -0,0 +1,250 @@
+#include <iostream>
+#include <memory>
+#include <utility>
+
+// This file contains codes used in the Spring2024 15445/645 C++ bootcamp.
```
Comment (abigalekim) [inline src/spring2024/s24_my_ptr.cpp:5]: NIT: codes => code (change for rest of files)

File: src/spring2024/s24_my_ptr.cpp [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+// It dives deeply into C++ new features like move constructor/assignment, move semantics, unique_ptr, shared_ptr,
+// wrapper class, etc., by implementing a simple version of unique_ptr from scratch.
+// **IMPORTANT**: please BEGIN reading from MAIN function!
+
+// Notice: it is just a simple version of std::unique_pointer<T>. The real implementation is much more complex!
```
Comment (abigalekim) [inline src/spring2024/s24_my_ptr.cpp:10]: What is a simple version of std::unique_ptr? I'm assuming you mean "implementation below" or "our implementation here" or something similar/

File: src/spring2024/s24_my_ptr.cpp [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    ptr_ = new T;
+    *ptr_ = 0;
+    std::cout << "New object on the heap: " << *ptr_ << std::endl;
+  }
+  Pointer(T val) {
+    ptr_ = new T;
+    *ptr_ = val;
+    std::cout << "New object on the heap: " << val << std::endl;
+  }
+  // DESTRUCTOR: Be called whenever an instance gets out of scope (just when the stack pops).
```
Comment (abigalekim) [inline src/spring2024/s24_my_ptr.cpp:25]: NIT: is called?

File: src/spring2024/s24_my_ptr.cpp [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+  ~Pointer() {
+    if (ptr_) {
+      std::cout << "Freed: " << *ptr_ << std::endl;
+      delete ptr_;
+    }
+  }
+
+  // Copy-init constructor is explicitely deleted.
```
Comment (abigalekim) [inline src/spring2024/s24_my_ptr.cpp:33]: NIT: explicitly

File: src/spring2024/s24_my_ptr.cpp [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+  Pointer(const Pointer<T> &) = delete;
+  // Copy-assign operator is explicitely deleted.
+  Pointer<T> &operator=(const Pointer<T> &) = delete;
+
+  // Add move constructor: useful when we need to EXTEND the lifetime of an object!
```
Comment (abigalekim) [inline src/spring2024/s24_my_ptr.cpp:38]: (Is this file meant to be read after move_constructors.cpp? If so, this is fine. If not, maybe (consult this other file for the meaning of lifetime)

Comment (abigalekim) [reply src/spring2024/s24_my_ptr.cpp:38]: Also this section should probably be read after the move semantics files given we go into move constructors and assignments here.

File: src/spring2024/s24_my_ptr.cpp [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+    }
+    if (ptr_) {  // We must free the existing pointer before overwriting it! Otherwise we LEAK!!
+      delete ptr_;
+    }
+    ptr_ = another.ptr_;
+    another.ptr_ = nullptr;  // NOTE: L14 avoids freeing nullptr during the destruction.
+    return *this;
+  }
+
+  // Overload operator *, to make the Pointer<T> feel like a "pointer", we can do this.
```
Comment (abigalekim) [inline src/spring2024/s24_my_ptr.cpp:53]: Note that the line below this is an example of the following syntax we can use with our unique ptr type.

File: src/spring2024/s24_my_ptr.cpp [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+
+void not_take_ownership(int *p) {
+  // Never `delete p` here!!
+}
+
+int main() {
+  /* ======================================================================
+     === Part 1: Common errors you come across in bustub ==================
+     ====================================================================== */
+  // You will see a variable type called "unique_ptr"...
```
Comment (abigalekim) [inline src/spring2024/s24_my_ptr.cpp:92]: When coding in C++/in this class, you will see...

File: src/spring2024/s24_my_ptr.cpp [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+  std::unique_ptr<int> ptr = std::make_unique<int>(3);
+  // What does this mean? Why we don't use the raw pointer like `int *p = new int`? (The answer is in Part 2)
+  // And later, when you need to pass this unique_ptr to a function, you may write the following code (please
+  // try to uncomment next line)...
+  // take_ownership(ptr);
+  // It doesn't work. The error is `Call to implicitly-deleted copy constructor of 'std::unique_ptr<int>'`.
+  // You may search in the Internet, and other people tell you to add a thing called `std::move`...
```
Comment (abigalekim) [inline src/spring2024/s24_my_ptr.cpp:99]: NIT: search the Internet

File: src/spring2024/s24_my_ptr.cpp [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+  // What's the problem of merely using the raw pointer?
+  int *p = new int;  // Malloc
+  *p = 456 * 12 / 34 + 23;
+  if (*p == 76) {
+    delete p;  // You may forget to add this line, and have `memory leak` problem!
+    return 0;
+  }
+  delete p;  // Free
+
+  // Raw pointers are dangerous! If you don't pay attention, you will come across problems like memory leak, double
```
Comment (abigalekim) [inline src/spring2024/s24_my_ptr.cpp:120]: NIT: memory leaks, double freeing

File: src/spring2024/s24_my_ptr.cpp [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+  // free, use after freeing...
+  // The reason is that in C++, raw pointers have no inherent mechanism to clean up automatically!
+  // Programmers have to malloc and free all by themselves, and it is easy to go wrong.
```
Comment (abigalekim) [inline src/spring2024/s24_my_ptr.cpp:123]: NIT: allocate and deallocate memory all by themselves

File: src/spring2024/s24_my_ptr.cpp [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+  /* ======================================================================
+     === Part 3: Let's implement a unique_ptr class from scratch ==========
+     ====================================================================== */
+  // We only show the last version of our own unique_ptr class.
+  // Here is the brief roadmap during the implementation:
+  //  1. First version: with default copy constructor & assignment, without move constructor & assignment
+  //     Problem: `Pointer p2 = p1` will cause `double free` problem
+  //     Copy constructor & assignment are evil in this case, since it will allow both p1 and p2 to manage the same raw
+  //     pointer!
+  //     Soluntion: disable copy constructor & assignment
```
Comment (abigalekim) [inline src/spring2024/s24_my_ptr.cpp:145]: NIT: misspelled solution

**Review by abigalekim — CHANGES_REQUESTED** [review]

File: src/spring2024/s24_my_ptr.cpp (hunk 8 above)
Comment (lanlou1554) [reply src/spring2024/s24_my_ptr.cpp:41]: Yeah! Good point! I add an important note in the beginning of this file.

**Comment by lanlou1554** [comment]
Thanks Abi!! I have fixed all the stuff and PTAL!

**Pull request merged**